### PR TITLE
LE Audio: Unicast group refactor

### LINKED ISSUE #45374: Creating the unicast group before both ISO connections have been configured might cause issue
**Describe the bug**
I try to modify the unicast_audio_client and unicast_audio_server for creating TWS scenario demo (one client connects to two servers).
And if the unicast group is created before both ISO connections have been configured, then the CIG creating parameter will be wrong.

I'm using controller from [here](https://github.com/nrfconnect/sdk-nrf/blob/v1.9.99-dev1/applications/nrf5340_audio/bin/ble5-ctr-rpmsg_3247.hex) with the application in https://github.com/rick1082/zephyr/tree/multilink_audio

And running the on nRF5340-DK.

**To Reproduce**
1. Checkout the unicast_audio_client from above branch and program to one nRF5340-DK as client
2. Checkout the unicast_audio_server from above branch, and program to nRF5340-DK as left headset server
3. Checkout the unicast_audio_server from above branch, remove following line and program to nRF5340-DK as right headset server
https://github.com/rick1082/zephyr/blob/multilink_audio/samples/bluetooth/unicast_audio_server/src/main.c#L60
4. Program all the DKs with ble5-ctr-rpmsg_3247.hex on the network core
5. Turn on both of servers, and then turn on the client
6. Then we see the client shows following log
> [00:00:10.104,827] <wrn> bt_hci_core: opcode 0x2064 status 0x0c
[00:00:10.104,858] <dbg> bt_iso: bt_iso_chan_connect: Failed to connect CISes: -5
[00:00:10.104,858] <err> bt_unicast_client: bt_audio_stream_connect: -5
7. And in the HCI traffic log, it shows following log indicated the CIG only created with one CIS
> < HCI Command: LE Set Connected I.. (0x08|0x0062) plen 24  #185 [hci0] 5.555816
        CIG ID: 0x00
        Master to Slave SDU Interval: 10000 us (0x002710)
        Slave to Master SDU Interval: 10000 us (0x002710)
        SCA: 201 - 500 ppm (0x00)
        Packing: Sequential (0x00)
        Framing: Unframed (0x00)
        Master to Slave Maximum Latency: 10 ms (0x000a)
        Slave to Master Maximum Latency: 10 ms (0x000a)
        Number of CIS: 1
        CIS ID: 0x00
        Master to Slave Maximum SDU Size: 40
        Slave to Master Maximum SDU Size: 0
        Master to Slave PHY: LE 2M (0x02)
        Slave to Master PHY: LE 2M (0x02)
        Master to Slave Retransmission attempts: 0x02
        Slave to Master Retransmission attempts: 0x00
> HCI Event: Command Complete (0x0e) plen 8                #186 [hci0] 5.559716
      LE Set Connected Isochronous Group Parameters (0x08|0x0062) ncmd 1
        Status: Success (0x00)
        CIG ID: 0x00
        Number of Handles: 1
        Connection Handle: 2

**Expected behavior**
The bt_audio API should work for creating the unicast group before both ISO connections been configured.

**Logs and console output**
The btmon format HCI log from client as following:
[client_hci.log](https://github.com/zephyrproject-rtos/zephyr/files/8630164/client_hci.log)
The application log from client as following
[client_application.log](https://github.com/zephyrproject-rtos/zephyr/files/8630195/client_application.log)

### LINKED ISSUE #45741: LE Audio: Allow unique `bt_codec_qos` for each unicast stream
**Is your enhancement proposal related to a problem? Please describe.**
The spec does not mandate that all streams in a unicast group for a specific unicast server shall use the same `bt_codec_qos` settings. 

The current API only supports supplying a single `bt_codec_qos` in the QoS function:
```c
int bt_audio_stream_qos(struct bt_conn *conn,
			struct bt_audio_unicast_group *group,
			struct bt_codec_qos *qos);
```

This should be made into an array, or somehow otherwise be attached to the individual streams. 

**Describe the solution you'd like**
Allow for more than a single `bt_codec_qos` for streams in a unicast group on a single server, i.e. different QoS settings for outgoing and incoming streams. 

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Thalley**
This PR refactors how the unicast group works in terms of when it is created, when it can be modified and how it is deleted. In a nutshell, the unicast group will now have the same lifetime as the underlying CIG. This gives us the same advantages and disadvantages of the CIG, and more tightly couples the unicast group to the ISO layer. 

This required significant rewrite of some of the underlying structures as ISO channels need to be allocated once the bt_audio_streams were added to the group, instead of when the streams were configured. 

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/45374
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/45741

### CODE REVIEW

**Comment by rick1082** [comment]
Hello @Thalley,
Thanks for the great PR, I did a quick test with modified unicast_audio_client which search for two unicast_audio_server, connected and then start ISO stream:
https://github.com/rick1082/zephyr/blob/dual_link_unicast_audio/samples/bluetooth/unicast_audio_client/src/main.c

For now I'm able to:
1. Create group, scan, connect, configure, set QoS and enable the two stream if both unicast_audio_servers are connected in short time.
But if calling bt_audio_stream_start() for both unicast_audio_servers in short time, the only one stream can start properly.
The root cause could be the controller from unicast_audio_client try to connect two CISes in short time without waiting for the first one finished.
The error log is like this:

> 00:00:04.868,011] <wrn> bt_hci_core: opcode 0x2064 status 0x0c                                                                                                                                                    
[00:00:04.868,011] <dbg> bt_iso: bt_iso_chan_connect: Failed to connect CISes: -5                                                                                                                                  
[00:00:04.868,041] <err> bt_unicast_client: bt_audio_stream_connect: -5                                                                                                                                            
[00:00:04.868,041] <dbg> bt_audio_stream: bt_audio_stream_start: Starting stream failed: -5                                                                                                                        
[00:00:04.868,041] <err> main: Unable to start stream: -5  

2. But if I let unicast_audio_client creates group for two streams, and then only connect to one unicast_audio_server
Then the unicast_audio_client will assert, the error log as following

> [00:00:00.289,581] <dbg> bt_audio_stream: bt_audio_unicast_group_create: Added stream 0x20001d68 to group 0x200022d4
[00:00:00.289,581] <dbg> bt_audio_stream: bt_audio_unicast_group_create: Added stream 0x20001d8c to group 0x200022d4
[00:00:00.289,611] <dbg> bt_audio_stream: bt_audio_cig_create: group 0x200022d4 qos 0x20000194
[00:00:00.289,642] <dbg> bt_iso: bt_iso_chan_add: iso 0x20001928 chan 0x200022f0
[00:00:00.289,703] <dbg> bt_iso: bt_iso_chan_add: iso 0x200019f0 chan 0x20002324
[00:00:00.290,252] <inf> main: Initialized
[00:00:00.290,252] <inf> main: Waiting for connection
[00:00:00.292,877] <inf> main: Scanning successfully started
[00:00:00.350,524] <inf> main: Connected: D7:A6:2B:3E:76:29 (random)
[00:00:00.353,332] <inf> main: Scanning successfully started
[00:00:02.600,067] <inf> main: Security changed: level 2
[00:00:02.850,921] <inf> main: discover sink cb for left
[00:00:03.350,189] <inf> main: discover sink cb for left
[00:00:03.350,219] <inf> main: Sink #0: ep 0x20002570
[00:00:04.851,531] <inf> main: discover sink cb for left
[00:00:04.851,562] <inf> main: Discover complete: err 0
[00:00:04.851,562] <dbg> bt_audio_stream: bt_audio_stream_config: conn 0x20001388 stream 0x20001d68, ep 0x20002570 codec 0x20000130 codec id 0x06 codec cid 0x0000 codec vid 0x0000
[00:00:04.851,593] <dbg> bt_audio_stream: bt_audio_stream_attach: conn 0x20001388 stream 0x20001d68 ep 0x20002570 codec 0x20000130
[00:00:04.851,654] <inf> main: configure stream for sink[HEADSET_L], err = 0
[00:00:05.725,250] <inf> main: Audio Stream 0x20001d68 configured, conn 0x20001388
[00:00:05.725,280] <dbg> bt_audio_stream: bt_audio_stream_qos: conn 0x20001388 group 0x200022d4
[00:00:05.725,280] <err> os: ***** MPU FAULT *****
[00:00:05.725,311] <err> os:   Data Access Violation
[00:00:05.725,311] <err> os:   MMFAR Address: 0xa0
[00:00:05.725,341] <err> os: r0/a1:  0x00000000  r1/a2:  0x20002324  r2/a3:  0x20001d8c
[00:00:05.725,341] <err> os: r3/a4:  0x20002324 r12/ip:  0x00000000 r14/lr:  0x0000f2d9
[00:00:05.725,372] <err> os:  xpsr:  0x61000000
[00:00:05.725,372] <err> os: s[ 0]:  0x00000000  s[ 1]:  0x00000000  s[ 2]:  0x00000000  s[ 3]:  0x00000000
[00:00:05.725,402] <err> os: s[ 4]:  0x00000000  s[ 5]:  0x00000000  s[ 6]:  0x00000000  s[ 7]:  0x00000000
[00:00:05.725,402] <err> os: s[ 8]:  0x00000000  s[ 9]:  0x00000000  s[10]:  0x00000000  s[11]:  0x00000000
[00:00:05.725,402] <err> os: s[12]:  0x00000000  s[13]:  0x00000000  s[14]:  0x00000000  s[15]:  0x00000000
[00:00:05.725,433] <err> os: fpscr:  0x00100000
[00:00:05.725,433] <err> os: Faulting instruction address (r15/pc): 0x0001965c
[00:00:05.725,463] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:05.725,463] <err> os: Current thread: 0x200010a0 (unknown)
[00:00:05.897,644] <err> os: Halting system


Faulting instruction is pointed to
https://github.com/Thalley/zephyr/blob/6494f1a8aa62b22bc57fed8eb184e437cf028cd8/subsys/bluetooth/audio/unicast_client.c#L233
LR is pointed to
https://github.com/Thalley/zephyr/blob/6494f1a8aa62b22bc57fed8eb184e437cf028cd8/include/zephyr/sys/slist.h#L274

**Comment by Thalley** [comment]
> Hello @Thalley, Thanks for the great PR, I did a quick test with modified unicast_audio_client which search for two unicast_audio_server, connected and then start ISO stream: https://github.com/rick1082/zephyr/blob/dual_link_unicast_audio/samples/bluetooth/unicast_audio_client/src/main.c
> 
> For now I'm able to:
> 
>     1. Create group, scan, connect, configure, set QoS and enable the two stream if both unicast_audio_servers are connected in short time.
>        But if calling bt_audio_stream_start() for both unicast_audio_servers in short time, the only one stream can start properly.
>        The root cause could be the controller from unicast_audio_client try to connect two CISes in short time without waiting for the first one finished.
>        The error log is like this:
> 
> 
> > 00:00:04.868,011]  bt_hci_core: opcode 0x2064 status 0x0c
> > [00:00:04.868,011]  bt_iso: bt_iso_chan_connect: Failed to connect CISes: -5
> > [00:00:04.868,041]  bt_unicast_client: bt_audio_stream_connect: -5
> > [00:00:04.868,041]  bt_audio_stream: bt_audio_stream_start: Starting stream failed: -5
> > [00:00:04.868,041]  main: Unable to start stream: -5
> 
>     2. But if I let unicast_audio_client creates group for two streams, and then only connect to one unicast_audio_server
>        Then the unicast_audio_client will assert, the error log as following
> 
> 
> > [00:00:00.289,581]  bt_audio_stream: bt_audio_unicast_group_create: Added stream 0x20001d68 to group 0x200022d4
> > [00:00:00.289,581]  bt_audio_stream: bt_audio_unicast_group_create: Added stream 0x20001d8c to group 0x200022d4
> > [00:00:00.289,611]  bt_audio_stream: bt_audio_cig_create: group 0x200022d4 qos 0x20000194
> > [00:00:00.289,642]  bt_iso: bt_iso_chan_add: iso 0x20001928 chan 0x200022f0
> > [00:00:00.289,703]  bt_iso: bt_iso_chan_add: iso 0x200019f0 chan 0x20002324
> > [00:00:00.290,252]  main: Initialized
> > [00:00:00.290,252]  main: Waiting for connection
> > [00:00:00.292,877]  main: Scanning successfully started
> > [00:00:00.350,524]  main: Connected: D7:A6:2B:3E:76:29 (random)
> > [00:00:00.353,332]  main: Scanning successfully started
> > [00:00:02.600,067]  main: Security changed: level 2
> > [00:00:02.850,921]  main: discover sink cb for left
> > [00:00:03.350,189]  main: discover sink cb for left
> > [00:00:03.350,219]  main: Sink #0: ep 0x20002570
> > [00:00:04.851,531]  main: discover sink cb for left
> > [00:00:04.851,562]  main: Discover complete: err 0
> > [00:00:04.851,562]  bt_audio_stream: bt_audio_stream_config: conn 0x20001388 stream 0x20001d68, ep 0x20002570 codec 0x20000130 codec id 0x06 codec cid 0x0000 codec vid 0x0000
> > [00:00:04.851,593]  bt_audio_stream: bt_audio_stream_attach: conn 0x20001388 stream 0x20001d68 ep 0x20002570 codec 0x20000130
> > [00:00:04.851,654]  main: configure stream for sink[HEADSET_L], err = 0
> > [00:00:05.725,250]  main: Audio Stream 0x20001d68 configured, conn 0x20001388
> > [00:00:05.725,280]  bt_audio_stream: bt_audio_stream_qos: conn 0x20001388 group 0x200022d4
> > [00:00:05.725,280]  os: ***** MPU FAULT *****
> > [00:00:05.725,311]  os:   Data Access Violation
> > [00:00:05.725,311]  os:   MMFAR Address: 0xa0
> > [00:00:05.725,341]  os: r0/a1:  0x00000000  r1/a2:  0x20002324  r2/a3:  0x20001d8c
> > [00:00:05.725,341]  os: r3/a4:  0x20002324 r12/ip:  0x00000000 r14/lr:  0x0000f2d9
> > [00:00:05.725,372]  os:  xpsr:  0x61000000
> > [00:00:05.725,372]  os: s[ 0]:  0x00000000  s[ 1]:  0x00000000  s[ 2]:  0x00000000  s[ 3]:  0x00000000
> > [00:00:05.725,402]  os: s[ 4]:  0x00000000  s[ 5]:  0x00000000  s[ 6]:  0x00000000  s[ 7]:  0x00000000
> > [00:00:05.725,402]  os: s[ 8]:  0x00000000  s[ 9]:  0x00000000  s[10]:  0x00000000  s[11]:  0x00000000
> > [00:00:05.725,402]  os: s[12]:  0x00000000  s[13]:  0x00000000  s[14]:  0x00000000  s[15]:  0x00000000
> > [00:00:05.725,433]  os: fpscr:  0x00100000
> > [00:00:05.725,433]  os: Faulting instruction address (r15/pc): 0x0001965c
> > [00:00:05.725,463]  os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
> > [00:00:05.725,463]  os: Current thread: 0x200010a0 (unknown)
> > [00:00:05.897,644]  os: Halting system
> 
> Faulting instruction is pointed to https://github.com/Thalley/zephyr/blob/6494f1a8aa62b22bc57fed8eb184e437cf028cd8/subsys/bluetooth/audio/unicast_client.c#L233 LR is pointed to https://github.com/Thalley/zephyr/blob/6494f1a8aa62b22bc57fed8eb184e437cf028cd8/include/zephyr/sys/slist.h#L274

Thanks, I'll look into this. Basically what is happening is that we are trying to connect multiple CIS in multiple commands at the same time, right?

**Comment by rick1082** [comment]
> 

Thank you @Thalley,
Yes, the item one is because the host should ask controller to finish the first CIS connection and then start the second one.
Otherwise only one CIS will be connected properly and the other one will fail.
But the item two is tricky, which leads to the scenario described in #45374 cannot be achieved.

**Comment by Thalley** [comment]
> > 
> 
> Thank you @Thalley, Yes, the item one is because the host should ask controller to finish the first CIS connection and then start the second one. Otherwise only one CIS will be connected properly and the other one will fail. But the item two is tricky, which leads to the scenario described in #45374 cannot be achieved.

OK, thanks. 

I'll try to see if I can find a solution for that, and somehow test it (the upstreamed unicast client sample does not support multiple servers right now).

**Comment by Thalley** [comment]
Marking as draft as this will be rebased on top of https://github.com/zephyrproject-rtos/zephyr/pull/46440

**Comment by Thalley** [comment]
> > 
> 
> Thank you @Thalley, Yes, the item one is because the host should ask controller to finish the first CIS connection and then start the second one. Otherwise only one CIS will be connected properly and the other one will fail. But the item two is tricky, which leads to the scenario described in #45374 cannot be achieved.

Created https://github.com/zephyrproject-rtos/zephyr/issues/46829 to deal with that (it is a separate issue to this PR).

**Comment by rick1082** [comment]
Confirmed that this fixed the #45374, and also the issue#2 which I described in the previous comment.
The issue#1 it not a show stopper, since we can add retry mechanism if we cannot establish the ISO channel properly.

File: samples/bluetooth/unicast_audio_client/src/main.c [hunk 1]
```diff
@@ -815,14 +834,16 @@ static int configure_streams(void)
 
 static int create_group(void)
 {
-	struct bt_audio_stream *streams_p[ARRAY_SIZE(streams)];
+	struct bt_audio_unicast_group_param params[ARRAY_SIZE(streams)];
 	int err;
 
 	for (size_t i = 0U; i < configured_stream_count; i++) {
-		streams_p[i] = &streams[i];
+		params[i].stream = &streams[i];
```
Comment (rick1082) [inline samples/bluetooth/unicast_audio_client/src/main.c:841]: This is nice! so now either the developer can create unicast_group with predefined settings or can fetch the settings during configuration.

**Review by rick1082 — DISMISSED** [review]

**Comment by Thalley** [comment]
Rebased to solve merge conflicts

**Comment by Thalley** [comment]
Rebased to solve merge conflicts

**Comment by Thalley** [comment]
Note to self: There seemingly is still an issue with the unicast client sample if the unicast server sample restarts; need to investigate and fix as part of this change

**Comment by Thalley** [comment]
Rebased to solve merge conflicts

**Review by asbjornsabo — COMMENTED** [review]
Not sure I am able to review this properly, but seems sane so far

File: subsys/bluetooth/audio/unicast_client.c [hunk 2]
```diff
@@ -47,8 +47,8 @@ static const struct bt_uuid *cp_uuid = BT_UUID_ASCS_ASE_CP;
 
 static struct bt_audio_ep snks[CONFIG_BT_MAX_CONN][CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SNK_COUNT];
 static struct bt_audio_ep srcs[CONFIG_BT_MAX_CONN][CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SRC_COUNT];
-static struct bt_audio_iso isos[CONFIG_BT_MAX_CONN][CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SNK_COUNT +
-						    CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SRC_COUNT];
+static struct bt_audio_iso audio_isos[CONFIG_BT_AUDIO_UNICAST_CLIENT_GROUP_COUNT]
```
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/audio/unicast_client.c:50]: It changed from
```
isos[CONFIG_BT_MAX_CONN][CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SNK_COUNT + CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SRC_COUNT];
```
to 
```
audio_isos[CONFIG_BT_AUDIO_UNICAST_CLIENT_GROUP_COUNT]				     [CONFIG_BT_AUDIO_UNICAST_CLIENT_GROUP_STREAM_COUNT];
```

It is still a 2D array (as it was before). The difference is that the number of ISO channels we can support does not directly correlate the number of audio streams (as 1 ISO channel be used for 2 streams).

**Comment by Thalley** [comment]
> Note to self: There seemingly is still an issue with the unicast client sample if the unicast server sample restarts; need to investigate and fix as part of this change

Unable to easily reproduce the issue with the current state of the PR. Will remove the DNM until issue has otherwise been reproduced.

**Comment by Thalley** [comment]
Rebased to solve merge conflicts

**Review by rick1082 — APPROVED** [review]

File: samples/bluetooth/unicast_audio_client/src/main.c [hunk 3]
```diff
@@ -667,12 +688,23 @@ static void disconnected(struct bt_conn *conn, uint8_t reason)
 	bt_conn_unref(default_conn);
 	default_conn = NULL;
 
-	start_scan();
```
Comment (asbjornsabo) [inline samples/bluetooth/unicast_audio_client/src/main.c:670]: Is it intentional to no longer start scanning again at disconnection?

Comment (asbjornsabo) [reply samples/bluetooth/unicast_audio_client/src/main.c:670]: Or will that be taken care of by the main loop?

**Review by asbjornsabo — APPROVED** [review]
LGTM

**Review by carlescufi — APPROVED** [review]

**Pull request merged**